monkey/agent_plugins/exploiters/ssh/src/ssh_client.py

```diff
@@ -27,36 +26,61 @@ def __init__(self):
     def connect(
         self, host: TargetHost, credentials: Credentials, port: NetworkPort, timeout: float
     ):
-        self._client = paramiko.SSHClient()
-        self._client.set_missing_host_key_policy(paramiko.WarningPolicy())
+        """
+        Connect to the host using SSH
+
+        Credentials may be a username and password, or a username and private key.
+
+        :param host: Host to connect to
+        :param credentials: Credentials to use for the connection
+        :param port: Port to connect to
+        :param timeout: Timeout for the connection, in seconds
+        :raises Exception: If the connection could not be established
+        """
+
+        if isinstance(credentials.secret, SSHKeypair):
+            connect_function = self._connect_with_private_key
+        elif isinstance(credentials.secret, Password):
+            connect_function = self._connect_with_login_credentials
+        else:
+            message = "Unrecognised credential secret type"
+            logger.debug(message)
+            raise Exception(message)
+
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(paramiko.WarningPolicy())
+
         try:
-            self._connect_with_private_key(host, credentials, port, timeout)
-            logger.debug("Successfully authentiated using SSH on host: {host.ip}")
-        except Exception:
-            try:
-                self._connect_with_login_credentials(host, credentials, port, timeout)
-            except Exception as err:
-                raise err
+            connect_function(client, host, credentials, port, timeout)
+            logger.debug(f"Successfully authenticated using SSH on host: {host.ip}")
+        except Exception as err:
+            client.close()
+            raise err
```

Comment:
```suggestion
            connect_function(client, host, credentials, port, timeout)
            logger.debug(f"Successfully authenticated using SSH on host: {host.ip}")
        finally:
            client.close()
```

Reply:
We don't want to close the client if the connection was successful. In that case, we're assigning it to `self._client` for further use.

Reply:
Got it.